File: 297SerializeandDeserializeBinaryTree.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <algorithm>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <list>
#include <sstream>

// #include "nestedVectorIntVisualization.h"
// #include "nestedVectorCharVisualization.h"
 #include "treeVisualizationinPreorder.h"
// #include "listNodeVisualization.h"

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (root == nullptr)
            return "";
        string res = "";
        queue<TreeNode*> nodeQueue({ root });
        
        while (!nodeQueue.empty())
        {
            TreeNode* temp = nodeQueue.front();
            nodeQueue.pop();
            if (temp != nullptr)
            {
                res += to_string(temp->val) + " ";
                nodeQueue.push(temp->left);
                nodeQueue.push(temp->right);
            }
            else
                res += "# ";
        }
        return res;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.empty())
            return nullptr;

        queue<TreeNode*> nodeQueue;
        // isstringstream: extract string from data
        istringstream is(data);
        string val;
        is >> val;
        TreeNode* res = new TreeNode(stoi(val));
        TreeNode* cur = res;

        nodeQueue.push(cur);

        while (!nodeQueue.empty())
        {
            TreeNode* temp = nodeQueue.front();
            nodeQueue.pop();
            is >> val;
            if (val.empty())
                break;
            if (val != "#")
            {
                cur = new TreeNode(stoi(val));
                nodeQueue.push(cur);
                temp->left = cur;
            }
            is >> val;
            if (val.empty())
                break;
            if (val != "#")
            {
                cur = new TreeNode(stoi(val));
                nodeQueue.push(cur);
                temp->right = cur;
            }
        }
        return res;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));


int main()
{
    TreeNode* root = new TreeNode(10);
    
    TreeNode* n11 = new TreeNode(2);
    TreeNode* n12 = new TreeNode(3);

    TreeNode* n21 = new TreeNode(4);
    TreeNode* n22 = new TreeNode(5);

    root->left = n11;
    root->right = n12;
    n12->left = n21;
    n12->right = n22;

    Codec codec;

    string res = codec.serialize(root);
    
    TreeNode* head = codec.deserialize(res);

    visualizeTreeInPreorder(head);
}

File: Python Solutions/297. Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
Use BFS and reverse-BFS for the approach.
"""

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # Encode is a standard BFS problem
        if root is None:
            return "#"
        nodeQueue = collections.deque([root])
        result = ""
        while nodeQueue:
            curNode = nodeQueue.popleft()
            if curNode is not None:
                result += str(curNode.val) + ","
                nodeQueue.append(curNode.left)
                nodeQueue.append(curNode.right)
            else:
                result += "#" + ","
        result = result[:-1]  # the last char is ",", so remove it
        return result

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        # Decode edge case
        if data == "#":
            return None
        # Decode standard case
        # start building the root node
        strArray = data.split(",")  # either a string of number or "#"
        # the 0-idex element must be a number because of edge case handling
        root = TreeNode(strArray[0])  
        # use a BFS to do parent node -> left child -> right child
        nodeQueue = collections.deque([root])
        # 0-index node has been created 
        idx = 1  
        while idx < len(strArray):
            # this node will be parent
            curNode = nodeQueue.popleft()  
            # Let's focus on left child
            leftVal = strArray[idx]
            if leftVal == "#":
                leftChild = None
            else:
                # don't forget to convert from str to int
                leftChild = TreeNode(int(leftVal))
                # This left child must be in queue because it will be parent for next layer
                nodeQueue.append(leftChild)
            # connect parent with left child
            curNode.left = leftChild
            idx += 1

            # focus on right child: repeat the same process
            rightVal = strArray[idx]
            if rightVal == "#":
                rightChild = None
            else:
                rightChild = TreeNode(int(rightVal))
                nodeQueue.append(rightChild)
            curNode.right = rightChild
            idx += 1
        return root
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))